rembed/stack.py

```diff
@@ -146,6 +148,71 @@ def _make_inputs(self):
         self.X = self.X or T.imatrix("X")
         self.transitions = self.transitions or T.imatrix("transitions")
 
+    def _step(self, transitions_t, ss_mask_gen_matrix_t, stack_t, buffer_cur_t, 
+            tracking_hidden_prev, stack_pushed, stack_merged, buffer):
+        batch_size, _ = self.X.shape
+        # Extract top buffer values.
+        idxs = buffer_cur_t + (T.arange(batch_size) * self.seq_length)
+        buffer_top_t = buffer[idxs]
+
+        if self._predict_network is not None:
+            # We are predicting our own stack operations.
+            predict_inp = T.concatenate(
+                [stack_t[:, 0], stack_t[:, 1], buffer_top_t], axis=1)
+            if self.use_tracking_lstm:
+                # update the hidden state and obtain predicted actions
+                tracking_hidden, actions_t = self._predict_network(
+                tracking_hidden_prev, predict_inp, self.model_dim * 3, 
+                self.tracking_lstm_hidden_dim, self._vs, name="predict_actions")
+            else:
+                # obtain predicted actions
+                actions_t = self._predict_network(
+                predict_inp, self.model_dim * 3, util.NUM_TRANSITION_TYPES, self._vs,
+                name="predict_actions")
+                # no change in hidden state because there is none
+                tracking_hidden = tracking_hidden_prev
+        else:
+            # no change in the hidden state because there is none
+            tracking_hidden = tracking_hidden_prev
+
+        if self.use_predictions: 
+            # Predicting our own actions
+            if self.interpolate:
+                # Interpolate between truth and prediction using bernoulli RVs 
+                # generated prior to the step
+                mask = (transitions_t * ss_mask_gen_matrix_t + 
```

Comment:
Hi @raghavgupta93, I saw the two definitions of the step function in combined them into one. I was also facing the problem with passing a None and hence changed ss_mask_gen_matrix_t to a dummy output variable which defaults to a tensor of size 1 when it is not used. This is much cheaper as compared to using a tensor of all 1s. Makes the code a bit dirty but helps us from defining two very similar step functions.